content/articles/haskell-websocket-tsung-benchmark.md

```diff
@@ -0,0 +1,275 @@
+that just broadcasts incoming messages to everyone.
+
+> **tsung** is a tcp/web load generator written in Erlang configured
+> through a XML domain-specific language
+> ([website](tsung.erlang-projects.org)).
+
+The core handler of our chat program looks like this (full source [here](https://github.com/WeAreWizards/haskell-websockets-tsung-benchmark/blob/master/code/src/Main.hs)):
+
+```haskell
+handleWS :: InChan ByteString -> PendingConnection -> IO ()
+handleWS bcast pending = do
+    localChan <- dupChan bcast
+    connection <- acceptRequest pending
+
+    -- Yes, we're leaking resources here :)
+    forkIO $ forever $ do
+        message <- readChan localChan
+        sendTextData connection message
+
+    -- loop forever
+    let loop = do
+            Text message <- receiveDataMessage connection
+            writeChan bcast message
+            loop
+    loop
+```
+
+To run the ec2 machines we're using
+[nixops](https://nixos.org/nixops/) which also does the spot-price
+bidding for us:
+
+```
+nixops create '<nix/test-setup.nix>'
+nixops deploy
+```
+
+(See
+[here](https://github.com/WeAreWizards/haskell-websockets-tsung-benchmark/blob/master/nix/test-setup.nix)
+for the full configuration including kernel tuning).
+
+## tsung setup
+
+Unfortunately I could not get the distributed tsung going: The
+distributed testing uses an Erlang function called `slave:start` which
+connects through SSH and spawns Erlang on the remote host. This failed
+for reasons I didn't have time to debug.
+
+But without the distributed testing there's a problem: A single server
+can only open ~65000 connections because ports are limited to 16
+bits. We want more connections though!
+
+Luckily tsung supports using multiple virtual IP addresses for a
+single network interface out of the box. We went to Amazon and clicked
+"Assign new IP" a few times to assign more private IPs to our tsung
+box.
+
+Now we associate the new IPs with our network interface:
+
+```shell
+ip addr add 172.31.23.115/20 dev eth0
+ip addr add 172.31.23.113/20 dev eth0
+ip addr add 172.31.23.114/20 dev eth0
+ip addr add 172.31.23.112/20 dev eth0
+ip addr add 172.31.18.80/20 dev eth0
+ip addr add 172.31.18.81/20 dev eth0
+ip addr add 172.31.18.82/20 dev eth0
+ip addr add 172.31.18.83/20 dev eth0
+```
+
+We have a
+[slightly different](https://github.com/WeAreWizards/haskell-websockets-tsung-benchmark/blob/master/code/src/tsung-conf.xml)
+tsung config from the Phoenix people which we copy to our tsung box:
+
+```console
+$ nixops scp --to tsung-1 code/src/tsung-conf.xml tsung-conf.xml
+code/src/tsung-conf.xml -> root@52.31.104.126:tsung-conf.xml
+tsung-conf.xml                                                  100% 1494     1.5KB/s   00:00
+```
+
+# Running tsung
+
+Our Nix config tuned the TCP stack and increased kernel limits, but we
+still need to run `ulimit` to make sure we're not hitting the default
+limit of 1024 file descriptors:
```

Comment:
It can but this is a one-off so I didn't add it.

nixos.org/nixos/manual/options.html#opt-security.pam.loginLimits